### PR TITLE
API: Cleaning `numpy/__init__.py` and main namespace - Part 5 [NEP 52]

### PR DESCRIPTION
**Pull request by mtsokol**
Relevant issue: https://github.com/numpy/numpy/issues/24306

Scope of the PR:
- Remove `compare_chararrays` as it comes from a legacy module. It's still available as `np.char.compare_chararrays`.
- ~~Remove~~ Deprecate `chararray` as it comes from a legacy module. It's still available as `np.char.chararray`.
- Remove `format_parser` as it's the only function from `rec` present in the main namespace. It's still available as `np.rec.format_parser`.
- ~~Remove `recarray`. It's still available as `np.rec.recarray`.~~

### CODE REVIEW

**Comment by ngoldbaum** [comment]
It looks like pandas needs to be updated for the doc build to pass.

It also looks like the repr for `recarray` needs to be updated to reflect its new canonical location:

```
File "/tmp/tmpraxm5dc8/doc/source/user/basics.rec.rst", line 699, in doc/source/user/basics.rec.rst
Failed example:
    type(recordarr.bar)
Expected:
    <class 'numpy.recarray'>
Got:
    <class 'numpy.core.records.recarray'>
```

File: numpy/char/__init__.pyi [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    isnumeric as isnumeric,
+    isdecimal as isdecimal,
+    array as array,
+    asarray as asarray,
+    compare_chararrays as compare_chararrays,
+    chararray as chararray
+)
+
+__all__: list[str]
+__path__: list[str]
```
Comment (BvB93) [inline numpy/char/__init__.pyi:58]:
```suggestion
```
Minor nit: `__path__` is already handled and special cased by type checkers these days.

Comment (mtsokol) [reply numpy/char/__init__.pyi]: Thank you for pointing that out!

**Comment by rgommers** [comment]
Can we actually do this? It looks to me like `np.recarray` and `np.chararray` have too much usage (~9,600 and ~2,500 hits in code search for me). So this may be more disruption than it's worth.

**Comment by mtsokol** [comment]
> Can we actually do this? It looks to me like `np.recarray` and `np.chararray` have too much usage (~9,600 and ~2,500 hits in code search for me). So this may be more disruption than it's worth.

@rgommers These are used but the change to move to `np.rec.recarray` and `np.char.chararray` is straightforward and backward compatible. But I agree it will cause some disruption.  

I think we have three options:
1. Keep the PR in the current form: `np.recarray` and `np.chararray` are moved to `np.rec` and `np.char`.
2. Revert changes and both names are available from `np.*` namespace and `np.rec.*` and `np.char.*` namespaces as they are right now in `main` (Drawback: this will break "each function in exactly one place" rule that we're trying to enforce).
3. Keep `np.recarray` and `np.chararray` but remove them from their `np.rec.*` and `np.char.*` namespaces (Does it make sense?) - I see that `np.rec.recarray` and `np.char.chararray` are way less popular (a hundred hits in search compared to thousands you mentioned).

What do you think about these options? Which should I go with?

**Comment by rgommers** [comment]
> (Drawback: this will break "each function in exactly one place" rule that we're trying to enforce).

Yes, that'd be a little unfortunate. However, I'm not sure it's really that black-and-white. For all functions/objects we should have one documented/preferred location, but for some other popular functions we've also decided to leave a compatibility shim in place.

For `chararray` I think @ngoldbaum had a preference to move that away from the main namespace to make space for the better new functionality (NEP 55)? Not sure though. Let's see what he thinks about that one.

For `recarray` we don't really have a replacement, so I think it's okay to leave it in the main namespace as the preferred location. I'm okay keeping both `np.recarray` and `np.rec.recarray`, since it's a single object only, so we could consider this one as a case where practicality beats purity.

**Comment by ngoldbaum** [comment]
> For chararray I think @ngoldbaum had a preference to move that away from the main namespace to make space for the better new functionality (NEP 55)? Not sure though. Let's see what he thinks about that one.

If leaving taking it out of the main namespace would be too disruptive, it can stay. That said, a user-visible deprecation warning in NumPy 2.0 with a plan to remove it in a few versions also makes sense I think.

**Comment by mtsokol** [comment]
Hi @rgommers @ngoldbaum,

I updated this PR according to your feedback:
- `np.recarray` is available again (also `np.rec.recarray` can be used),
- `np.chararray` is available again but accessing it yields a `DeprecationWarning`.

File: doc/release/upcoming_changes/24587.python_removal.rst [hunk 2]
```diff
@@ -0,0 +1,8 @@
+* ``np.compare_chararrays`` has been removed from the main namespace. 
+  Use ``np.char.compare_chararrays`` instead.
+
+* ``np.chararray`` has been deprecated and it will be removed in the future.
+  Use ``np.char.chararray`` instead.
```
Comment (ngoldbaum) [inline doc/release/upcoming_changes/24587.python_removal.rst:5]: I think I would rephrase this slightly as "The `charrarray` in the main namespace has been deprecated. It can be imported without a deprecation warning from `np.char.chararray` for now, but we are planning to fully deprecate and remove `chararray` in the future.", since I think the intention is to fully deprecate it and remove it, we just haven't actually done the work yet, so can't outright remove it now. I proposed doing those deprecations before NumPy 2.0 as part of NEP 55, but that hasn't fully gone through yet. If that happens before NumPy 2.0, I will update this release note.

File: numpy/__init__.py [hunk 3]
```diff
@@ -385,6 +390,14 @@ def __getattr__(attr):
                 f"{__expired_attributes__[attr]}"
             )
 
+        if attr == "chararray":
+            warnings.warn(
+                "`np.chararray` is deprecated and will be removed from "
+                "the main namespace in the future. Use `np.char.chararray` "
+                "instead.", DeprecationWarning, stacklevel=2)
```
Comment (ngoldbaum) [inline numpy/__init__.py:397]: For the migration, I would say "Use an array with a string or bytes dtype instead," while it will still be available as `np.char.charray`, that will likely also be deprecated before NumPy 2.0 is released.

**Review by ngoldbaum — COMMENTED** [review]
While we may not be deprecating `np.recarray`, I think it would still make sense to change its repr, in this PR I have:

```
In [3]: np.recarray
Out[3]: numpy.recarray
```

Whereas I would expect it to be an alias to `numpy.rec.recarray` now rather than the other way around.

File: numpy/__init__.py (hunk 3 above)
Comment (mtsokol) [reply numpy/__init__.py]: Sure! Updated.

**Comment by mtsokol** [comment]
> While we may not be deprecating `np.recarray`, I think it would still make sense to change its repr, in this PR I have:
> 
> ```
> In [3]: np.recarray
> Out[3]: numpy.recarray
> ```
> 
> Whereas I would expect it to be an alias to `numpy.rec.recarray` now rather than the other way around.

@ngoldbaum I decided to keep repr as `numpy.recarray` - in my understanding all items present in the main namespace should have `numpy.*` repr. There is [a separate test](https://github.com/numpy/numpy/blob/4f84d719657eb455a35fcdf9e75b83eb1f97024a/numpy/tests/test_public_api.py#L20): `test_numpy_namespace` with `check_dir` function, that checks if main namespace functions have correct (`numpy.*`) `__module__` attribute. 

Should I change `recarray` repr and add to the `allowlist` in the test?

**Comment by ngoldbaum** [comment]
I personally think changing the canonical location makes sense and is in the spirit of the other changes we're making for the API migration. Feel free to wait for others to weigh in.

**Comment by mtsokol** [comment]
> I personally think changing the canonical location makes sense and is in the spirit of the other changes we're making for the API migration. Feel free to wait for others to weigh in.

Makes sense to me - I updated the repr to `numpy.rec.recarray`.

**Comment by ngoldbaum** [comment]
Thanks so much for your diligence on pushing forward refactoring the main namespace @mtsokol!
